backend/app/notify/email.go

```diff
@@ -0,0 +1,242 @@
+	"io"
+	"net"
+	"net/smtp"
+	"sync"
+	"text/template"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+	"github.com/go-pkgz/repeater"
+	"github.com/pkg/errors"
+)
+
+// EmailParams contain settings for email set up
+type EmailParams struct {
+	Host     string        // SMTP host
+	Port     int           // SMTP port
+	TLS      bool          // TLS auth
+	From     string        // From email field
+	Username string        // user name
+	Password string        // password
+	TimeOut  time.Duration // TLS connection timeout
+	Template string        // request message template
+}
+
+// Email implements notify.Destination for email
+type Email struct {
+	EmailParams
+	SMTPClient
+	template  *template.Template // request message template
+	sendMutex sync.Mutex         // Send is synchronious and blocked by this mutex
+	count     int                // amount of waiting + running send requests
+}
+
+// SMTPClient interface defines subset of net/smtp used by email client
+type SMTPClient interface {
+	Mail(string) error
+	Auth(smtp.Auth) error
+	Rcpt(string) error
+	Data() (io.WriteCloser, error)
+	Quit() error
+	Close() error
+}
+
+// tmplData store data for message from request template execution
+type tmplData struct {
+	From      string
+	To        string
+	Orig      string
+	Link      string
+	PostTitle string
+}
+
+const emailConnectionTimeOut = 10 * time.Second
+
+//NewEmail makes email object for notifications and run sending daemon
+func NewEmail(params EmailParams) (*Email, error) {
+
+	res := Email{EmailParams: params}
+	if res.TimeOut == 0 {
+		res.TimeOut = emailConnectionTimeOut
+	}
+	log.Printf("[DEBUG] create new email notifier for server %s with user %s, timeout=%s",
+		res.Host, res.Username, res.TimeOut)
+
+	var err error
+
+	tmpl := msgTemplate
+	if params.Template != "" {
+		tmpl = params.Template
+	}
+	res.template, err = template.New("messageFromRequest").Parse(tmpl)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse message template")
+	}
+	// establish test connection and
+	testSMTPClient, err := res.client()
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't establish test connection")
+	}
+	if err = testSMTPClient.Quit(); err != nil {
+		log.Printf("[WARN] failed to send quit command to %s:%d, %v", res.Host, res.Port, err)
+	}
+	if err := testSMTPClient.Close(); err != nil {
+		log.Printf("[WARN] can't close smtp connection, %v", err)
+	}
+	return &res, err
+}
+
+// Send email from request to address in settings
+func (e *Email) Send(ctx context.Context, req request) error {
+	log.Printf("[DEBUG] send notification via %s, comment id %s", e, req.comment.ID)
+	// TODO: decide where to get "to" email from
+	to := "test@localhost"
+	msg := e.buildMessageFromRequest(req, to)
+	err := repeater.NewDefault(5, time.Millisecond*250).Do(ctx, func() error {
+		return e.sendEmail(msg, to)
+	})
+	return err
+}
+
+// sendEmail sends message, prepared by Email.buildMessage
+func (e *Email) sendEmail(message, to string) error {
```

Comment:
Generally it's not clear to me what sendMail is doing, epscial the use of count. If there is a legit reason for this logic it should be expeined in the top comment to the function. However reading it I fail to see how this count increments/decerements help to achive the goal. I gues the goal is to queue it somehow, but not sure